File: packages/graph/src/serialize.ts
import { canonicalJson } from "@evidenceops/contracts";
import type { EvidenceGraph, GraphEdge, GraphNode, JsonValue } from "./types.js";

function byId<T extends { id: string }>(items: T[]): T[] {
  return [...items].sort((left, right) => left.id.localeCompare(right.id));
}

function asJson(value: unknown): JsonValue {
  return JSON.parse(JSON.stringify(value)) as JsonValue;
}

export function normalizeGraph(graph: EvidenceGraph): EvidenceGraph {
  return {
    schemaVersion: graph.schemaVersion,
    nodes: byId(graph.nodes),
    edges: byId(graph.edges),
  };
}

export function graphToCanonicalJson(graph: EvidenceGraph): string {
  return canonicalJson(asJson(normalizeGraph(graph)));
}

export function graphToJsonl(graph: EvidenceGraph): string {
  const normalized = normalizeGraph(graph);
  const lines = [`{"kind":"meta","schemaVersion":${normalized.schemaVersion}}`];
  for (const node of normalized.nodes) {
    lines.push(canonicalJson({ kind: "node", ...asJson(node) as Record<string, JsonValue> }));
  }
  for (const edge of normalized.edges) {
    lines.push(canonicalJson({ kind: "edge", ...asJson(edge) as Record<string, JsonValue> }));
  }
  return `${lines.join("\n")}\n`;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

export function graphFromJson(text: string): EvidenceGraph {
  const parsed: unknown = JSON.parse(text);
  if (!isRecord(parsed) || !Array.isArray(parsed.nodes) || !Array.isArray(parsed.edges)) {
    throw new Error("graph json: expected {schemaVersion,nodes,edges}");
  }
  return {
    schemaVersion: Number(parsed.schemaVersion),
    nodes: parsed.nodes as GraphNode[],
    edges: parsed.edges as GraphEdge[],
  };
}

export function graphFromJsonl(text: string): EvidenceGraph {
  const nodes: GraphNode[] = [];
  const edges: GraphEdge[] = [];
  let schemaVersion = 1;
  for (const line of text.split("\n")) {
    if (line.trim() === "") {
      continue;
    }
    const parsed: unknown = JSON.parse(line);
    if (!isRecord(parsed) || typeof parsed.kind !== "string") {
      throw new Error("graph jsonl: each line needs kind");
    }
    if (parsed.kind === "meta") {
      schemaVersion = Number(parsed.schemaVersion);
    } else if (parsed.kind === "node") {
      const { kind: _kind, ...node } = parsed;
      nodes.push(node as GraphNode);
    } else if (parsed.kind === "edge") {
      const { kind: _kind, ...edge } = parsed;
      edges.push(edge as GraphEdge);
    } else {
      throw new Error(`graph jsonl: unknown kind ${parsed.kind}`);
    }
  }
  return { schemaVersion, nodes, edges };
}

export function redactGraph(graph: EvidenceGraph): EvidenceGraph {
  return {
    schemaVersion: graph.schemaVersion,
    nodes: graph.nodes.map((node) => {
      if (node.redaction === "confidential" || node.redaction === "privileged") {
        return {
          ...node,
          sourceLocator: null,
          attributes: { redacted: true },
        };
      }
      return node;
    }),
    edges: graph.edges,
  };
}

File: src/evidenceops/graph/serialize.py
from __future__ import annotations

import json
from collections.abc import Mapping
from typing import Any

from evidenceops.contracts import canonical_json


def _sort_graph(graph: Mapping[str, Any]) -> dict[str, Any]:
    nodes = sorted(list(graph.get("nodes", [])), key=lambda node: str(node.get("id")))
    edges = sorted(list(graph.get("edges", [])), key=lambda edge: str(edge.get("id")))
    return {
        "schemaVersion": graph.get("schemaVersion"),
        "nodes": nodes,
        "edges": edges,
    }


def graph_to_canonical_json(graph: Mapping[str, Any]) -> str:
    return canonical_json(_sort_graph(graph))


def graph_to_jsonl(graph: Mapping[str, Any]) -> str:
    normalized = _sort_graph(graph)
    lines = [canonical_json({"kind": "meta", "schemaVersion": normalized["schemaVersion"]})]
    for node in normalized["nodes"]:
        payload = {"kind": "node", **node}
        lines.append(canonical_json(payload))
    for edge in normalized["edges"]:
        payload = {"kind": "edge", **edge}
        lines.append(canonical_json(payload))
    return "\n".join(lines) + "\n"


def graph_from_jsonl(text: str) -> dict[str, Any]:
    nodes: list[Any] = []
    edges: list[Any] = []
    schema_version = 1
    for line in text.split("\n"):
        if not line.strip():
            continue
        parsed = json.loads(line)
        kind = parsed.pop("kind", None)
        if kind == "meta":
            schema_version = int(parsed.get("schemaVersion", 1))
        elif kind == "node":
            nodes.append(parsed)
        elif kind == "edge":
            edges.append(parsed)
        else:
            raise ValueError(f"unknown jsonl kind {kind}")
    return {"schemaVersion": schema_version, "nodes": nodes, "edges": edges}
